src/misc.cpp

```diff
@@ -58,28 +67,61 @@ void InitializeCheats();
 void InitializeNPF();
 void InitializeOldNames();
 
+void RandomBytesWithFallback(std::span<uint8_t> buf)
+{
+#if defined(_WIN32)
+	auto res = BCryptGenRandom(nullptr, static_cast<PUCHAR>(buf.data()), static_cast<ULONG>(buf.size()), BCRYPT_USE_SYSTEM_PREFERRED_RNG);
+	if (res >= 0) return;
+#elif defined(__APPLE__) || defined(__NetBSD__) || defined(__FreeBSD__)
+	 arc4random_buf(buf.data(), buf.size());
+#elif defined(__GLIBC__) && ((__GLIBC__ > 2) || ((__GLIBC__ == 2) && (__GLIBC_MINOR__ >= 25)))
+	auto res = getrandom(buf.data(), buf.size(), 0);
+	if (res > 0 && static_cast<size_t>(res) == buf.size()) return;
+#elif defined(__EMSCRIPTEN__)
+	auto res = EM_ASM_INT({
+		var buf = $0;
+		var bytes = $1;
+
+		var crypto = window.crypto;
+		if (crypto === undefined || crypto.getRandomValues === undefined) {
+			return -1;
+		}
+
+		crypto.getRandomValues(Module.HEAPU8.subarray(buf, buf + bytes));
+		return 1;
+	}, buf.data(), buf.size());
+	if (res > 0) return;
+#else
+#	warning "No high quality random source available; using a fallback instead"
+#endif
+
+	Debug(net, 0, "High quality random source unavailable; using fallback");
+	for (uint i = 0; i < buf.size(); i++) {
+		buf[i] = static_cast<uint8_t>(InteractiveRandom());
+	}
+}
+
 /**
  * Generate an unique ID.
  *
- * It isn't as much of an unique ID as we would like, but our random generator
- * can only produce 32bit random numbers.
- * That is why we combine InteractiveRandom with the current (steady) clock.
- * The first to add a bit of randomness, the second to ensure you can't get
- * the same unique ID when you run it twice from the same state at different
- * times.
- *
- * This makes it unlikely that two users generate the same ID for different
- * subjects. But as this is not an UUID, so it can't be ruled out either.
+ * It isn't as much of an unique ID but more a hashed digest of a random
+ * string and a time. It is very likely to be unique, but it does not follow
+ * any UUID standard.
  */
 std::string GenerateUid(std::string_view subject)
 {
-	auto current_time = std::chrono::duration_cast<std::chrono::nanoseconds>(std::chrono::steady_clock::now().time_since_epoch()).count();
-	std::string coding_string = fmt::format("{}{}{}", InteractiveRandom(), current_time, subject);
+	uint8_t random_bytes[32] = {};
```

Comment:
What about the more C++ way of doing this:
```suggestion
	std::array<uint8_t, 32> random_bytes = {};
```
and `crypto_blake2b_update(&ctx, random_bytes.data(), random_bytes.size());` later on?

Reply:
Does that actually matter? All I see is more letters to type, and both are used in the C++ world. Is there a benefit from one over the other?

Reply:
It's mostly that raw arrays can lead to nasty effects when being passed as parameters; see https://wiki.sei.cmu.edu/confluence/display/c/ARR01-C.+Do+not+apply+the+sizeof+operator+to+a+pointer+when+taking+the+size+of+an+array
https://stackoverflow.com/a/77106115 also lists quite a number of benefits of `std::array` over raw array.

I'd be in favour of using `std::array`/`std::span` over raw arrays, but it'll definitely take a while before the code base has migrated to that. Though, it's something to strive for I'd say.

Reply:
You mentioned it lists "quite a number", but I actually can't find any? The things mentioned there we already mitigate in our codebase (for example via `lengthof` or are not applicable in this case?

To be clear, I already changed the things into `std::array`, but I am mostly curious to why it would be better, if there is any fundamental reason. For things like `static_cast` I know there are, but here I have a hard time finding the benefit of one over the other.

For `std::span` it btw is clear: keeping ptr + len combined together, that is an obvious improvement for a wide set of reasons. But for `std::array` I have a hard time seeing them; and I like to understand, not blindly follow :)

Reply:
I think https://stackoverflow.com/a/77106115 is the first place I managed to find which actually lists a set of benefits. Which are interesting. Most do not apply to this situation ofc, and mostly have to do with passing an array around (which we already do in spans), and how to initialize them (which is irrelevant here).

The one that stands out is:

```
int data[] = {0, 1};
int copy[] = data;
```

Which is invalid. But works with `std::array`. And that it doesn't allow VLA, which is a positive thing.